scenario/constraints/UtilizationConstraint.ts

```diff
@@ -107,43 +107,51 @@ export class UtilizationConstraint<T extends CometContext, R extends Requirement
-            borrowCollateralFactor: borrowCollateralFactor.toString(),
-            collateralNeeded: collateralNeeded.toString(),
-          };
-
-          let borrowActor = await context.allocateActor('UtilizationConstraint{Borrower}', info);
-
-          await context.sourceTokens(collateralNeeded, collateralToken, borrowActor);
-          await collateralToken.approve(borrowActor, comet);
-          await borrowActor.safeSupplyAsset({ asset: collateralToken.address, amount: collateralNeeded });
-
-          // XXX will also need to make sure there are enough base tokens in the protocol to withdraw
-          await comet.connect(borrowActor.signer).withdraw(baseToken.address, toBorrowBase);
+          // To borrow as much, we need to supply some collateral.
+          const numAssets = await comet.numAssets();
+          for (let i = 0; i < numAssets; i++) {
+            console.log(`UtilizatioConstraint: attempting to source from $asset${i}`);
+
+            const { asset: collateralAsset, borrowCollateralFactor, priceFeed, scale } = await comet.getAssetInfo(i);
+
+            const collateralToken = context.getAssetByAddress(collateralAsset);
+
+            const basePrice = (await comet.getPrice(await comet.baseTokenPriceFeed())).toBigInt();
+            const collateralPrice = (await comet.getPrice(priceFeed)).toBigInt();
+
+            const baseScale = (await comet.baseScale()).toBigInt();
+            const collateralScale = scale.toBigInt();
+
+            const collateralWeiPerUnitBase = (collateralScale * basePrice) / collateralPrice;
+            let collateralNeeded = (collateralWeiPerUnitBase * toBorrowBase) / baseScale;
+            collateralNeeded = (collateralNeeded * factorScale) / borrowCollateralFactor.toBigInt(); // adjust for borrowCollateralFactor
+            collateralNeeded = (collateralNeeded * 11n) / 10n; // add fudge factor
+
+            const info = {
+              utilization,
+              totalSupplyBase: totalSupplyBase.toString(),
+              totalBorrowBase: totalBorrowBase.toString(),
+              toBorrowBase: toBorrowBase.toString(),
+              collateralAsset,
+              borrowCollateralFactor: borrowCollateralFactor.toString(),
+              collateralNeeded: collateralNeeded.toString(),
+            };
+
+            const borrowActor = await context.allocateActor('UtilizationConstraint{Borrower}', info);
+
+            try {
+              await context.sourceTokens(collateralNeeded, collateralToken, borrowActor);
+              await collateralToken.approve(borrowActor, comet);
+              await borrowActor.safeSupplyAsset({ asset: collateralToken.address, amount: collateralNeeded });
+
+              // XXX will also need to make sure there are enough base tokens in the protocol to withdraw
+              await comet.connect(borrowActor.signer).withdraw(baseToken.address, toBorrowBase);
+              console.log(`UtilizatioConstraint: successfully sourced from $asset${i}`);
```

Comment:
Utilizatio YO

Reply:
it's the cool new way to say `Utilization`

Reply:
updated